js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -395,6 +647,101 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
     }
 }
 
+/**
+ * Responsible for handling the batching of changes for the purposes of calling listeners
+ */
+const defaultListenerBatcher = {
+    /**
+     * Whether batch is in progress
+     */
+    isActive: false,
+    /**
+     * This is items that need to be notified only once per change in a batch
+     */
+    pending: new Map(),
+    /**
+     * This is items that don't need to respect batching rules. This is really for internal
+     * use but is part of the public API. This exists as to make listeners work across
+     * related fields we listen to changes on the related field cache
+     */
+    pendingNoBatch: new Map(),
+    /**
+     * This is for callbacks that only listen to all changes on a model (not specific records or fields)
+     */
+    pendingAll: new Set(),
+
+    /**
+     * Queue a call to a non-record/field specific listeners
+     */
+    callAll(listener: AllChangesListener): void {
+        if (this.isActive) {
+            this.pendingAll.add(listener);
+        } else {
+            listener();
+        }
+    },
+    /**
+     * Queue a call to a record/field specific listener
+     *
+     * If `shouldBatch` is false then there's no guarantees about the listener only being called
+     * once.
+     */
+    call<T>(
+        listener: ChangeListener<T>,
+        before?: T | null,
+        after?: T | null,
+        shouldBatch = true
+    ): void {
+        if (this.isActive) {
+            const p = shouldBatch ? this.pending : this.pendingNoBatch;
+            let listenerPending = p.get(listener);
+            if (!listenerPending) {
+                p.set(listener, [before, after]);
+            } else {
+                listenerPending[1] = after;
+            }
+        } else {
+            listener(before, after);
+        }
+    },
+    /**
+     * Start a batch. The passed function is called and any changes queued. Once function returns listeners
+     * will be dispatched. The value returned from `run` will be returned.
+     *
+     * If any error occurs in `run` then no listeners will be called.
+     */
+    batch<T>(run: () => T): T {
+        if (this.isActive) {
+            return run();
+        }
+        this.isActive = true;
+        try {
+            const r = run();
+            let pending = this.pendingNoBatch;
+            // TODO: :shrug:?
+            while (pending.size > 0) {
+                this.pendingNoBatch = new Map();
+                pending.forEach(([before, after], cb) => cb(before, after));
+                pending = this.pendingNoBatch;
+                // break;
+            }
+            this.pending.forEach(([before, after], cb) => cb(before, after));
+            this.pending = new Map();
+            this.pendingAll.forEach(cb => cb());
+            this.pendingAll = new Set();
+            // TODO:Should this move up above this.pending.forEach?
```

Comment:
Why would you want to do this? Until this point isn't it still active?

Reply:
This is the 'flush' phase... if any of this:

```
this.pending.forEach(([before, after], cb) => cb(before, after));
            this.pending = new Map();
            this.pendingAll.forEach(cb => cb());
            this.pendingAll = new Set();
```

triggered another callback... then they would be batched but lost as we set `this.pending = new Map()` afterwards.

I wasn't sure what I should do here but I think in practice it won't happen (it's all internal stuff - I added `this.pendingNoBatch ` purely for this reason)

Reply:
Can you expand that comment then, eg:
```
// TODO:Should this move up above this.pending.forEach?
```
to
```
// TODO:Should this move up above this.pending.forEach?
// (If any of the cb() calls trigger another batch then those changes will trigger
// an error because `pendingNoBatch` is set so that we don't lose any changes)
```
(Assuming what I just wrote is correct)

Reply:
In latest commit I've removed this comment entirely with my changes to how listeners work